src/_zkapauthorizer/tests/test_replicate.py

```diff
@@ -318,240 +327,284 @@ def match(self, matchee: tuple[str, Callable[[], BinaryIO]]) -> Optional[Mismatc
+def repeatUntil(condition: Callable[[], bool], action: Callable[[], object]) -> None:
+    """
+    Run an action repeatedly until a condition is true.
+    """
+    while True:
+        action()
+        if condition():
+            break
 
-        tvs = self.useFixture(TemporaryVoucherStore(get_config, aware_now))
 
-        rwcap_file = FilePath(tvs.config.get_private_path(REPLICA_RWCAP_BASENAME))
-        rwcap_file.parent().makedirs()
-        rwcap_file.setContent(b"URL:DIR2:stuff")
+def is_event_stream(grid: MemoryGrid, **kwargs: Matcher) -> Matcher[tuple[str, dict]]:
+    """
+    Match a Tahoe-LAFS directory entry representing a file which can be
+    retrieved from the given grid and which contains an ``EventStream`` with a
+    structure matched by the given keyword arguments.
+    """
 
-        # we use this to contol when the first upload happens, so that we
-        # actually use the queue
-        wait_d: Deferred[None] = Deferred()
+    def is_filenode():
+        return AfterPreprocessing(lambda item: item[0], Equals("filenode"))
 
-        uploads = []
-        upload_completed = False
+    def download_event_stream(cap):
+        return EventStream.from_bytes(BytesIO(grid.download(cap)))
 
-        async def uploader(name, get_data):
-            uploads.append((name, get_data))
-            await wait_d
-            nonlocal upload_completed
-            upload_completed = True
+    return MatchesAll(
+        is_filenode(),
+        AfterPreprocessing(
+            lambda item: download_event_stream(item[1]["ro_uri"]),
+            MatchesStructure(**kwargs),
+        ),
+    )
 
-        async def pruner(predicate):
-            pass
 
-        srv = replication_service(tvs.store._connection, uploader, pruner)
+class ReplicationServiceTests(TestCase):
+    """
+    Tests for ``_ReplicationService``.
+    """
 
-        # run the service and produce some fake voucher etc changes
-        # that cause "events" to be issued into the database
-        srv.startService()
-        self.addCleanup(srv.stopService)
+    def test_enable_replication_on_connection(self) -> None:
+        """
+        When the service starts it enables replication on its database connection.
+        """
+        tvs = self.useFixture(TemporaryVoucherStore(aware_now))
+        store = tvs.store
 
-        def add_tokens():
-            tokens = [RandomToken(b64encode(urandom(96))) for _ in range(10)]
-            voucher = urlsafe_b64encode(urandom(32))
-            tvs.store.add(voucher, len(tokens), 1, lambda: tokens)
+        # We'll spy on "events" which are only captured when the connection is
+        # in replication mode.  To start, make sure that database changes are
+        # not already being captured.  They should not be since nothing has
+        # placed the connection into replication mode yet.
+        # store.start_lease_maintenance().finish()
+        # self.assertThat(get_events(store._connection).changes, HasLength(0))
```

Comment:
not clear if you meant to include these assertions (but I guess: delete them if not)

Reply:
(looks like they happen below, so "delete"?)

Reply:
Oops.  Good thing the test suite still passes with these in.  The ones below invert the condition (`Not(HasLength(0))`).  These are just meant to prove some assumptions the test is making holds (that there are no surprise changes in the database before the test even tries to do anything).